README.md
# 2D Median Filtering Module with VBUS interface

<!--
## Table of Contents
 1. [Overview](#Overview)
 2. [Applications](#Applications)
 3. [Resource Utilization](#Resource)
 4. [Operation](#Operation)
 5. [Parameters](#Parameters)
 6. [Pinout](#Pinout)
 7. [Interfaces](#Interfaces)
 8. [Source Files](#Source)
 9. [Padding Options](#Padding)
 10. [Simulation](#Simulation)
-->


## Overview
A 2D median filtering implementation in Verilog.
* Synthesizable, technology independent Verilog IP Core
* Apply median filtering to input stream
* All signals synchronous with the input clock
* Supports kernel size 3x3 fully; larger size kernels (e.g. 5x5) provides rough sorting !!!
* Supports masking; apply filtering only to non-mask regions
* Supports padding zeros/ones to regions outside image
* FPGA proven (Tested on XC7A100T)
</br>

## Applications
* Noise Filtering
* Salt and pepper noise removal
* Dead/bad pixel correction
and many more...
</br>

## Resource
##### Window Size = 3x3
|Part|Data Width|LUTs|Registers |BRAM|
|-|-:|-|-|-|
|Lattice Crosslink NX (LIFCL-40)|8|1203 (3.73 %)|372 (1.13%)|15 EBR (17.0 %)
|Xilinx Artix 7 (XC7A100T)|8|1055 (1.66%)|971 (0.77%)|7.5 BRAM (0.77%)

</br>

## Operation
#### *Block Diagram*
<img src="doc/img/blockdiagram.png" width="800">
</br>

**Figure 1.** The block diagram above describes the operation of this median filtering module.

#### *Programming*
Everything starts with programming the registers for WIDTH and HEIGHT parameters through slave memory bus interface. This operation causes an internal reset pulse generation. Additionally, EN bit should also be set to 1.

#### *Buffering*
If a valid video stream is given to slave video bus of the module, the pixels will be received to line buffers shown in the yellow block.

#### *Sorting and Median Finding*
When enough pixels are buffered, pixel stacks (one pixel from each line buffer) will be fed to sorting pipeline block shown in blue color. This module is responsible for finding the median from the pixels in the current window.

|Important Note|
|:-|
|The sorting block outputs the median value only for 3x3 window size.</br> For sizes larger than 3x3, the module performs a rough sorting. Consequently, a value close-neighbouring the median is given instead of exact median. This does not degrade the performance of filtering so much contrary to what is generally believed. |

#### *Output buffers*
Output buffering module gets processed lines from sorting block into a FIFO. These lines are then output through master video bus interface.

## Parameters
|Parameter Name|Description|Type|Valid Range|
|-|-|-|-|
|DW_VD        |Data width        |    integer    |DW_VD>1|
|SIZE        |Kernel size  </br> Note: The case where SIZE=3 performs full median filtering. </br> When set to sizes > 3, the module performs partial sorting and does not guarantee finding the exact medians. |integer    | 3, 5, 7, …. |
|IN_BUF_LEN    |Input buffer depth|integer|Powers of two|
|OUT_BUF_LEN|Output buffer depth|Powers of two|
|USE_VENDOR_FIFO|This parameter selects the FIFO instantiation source in the design.</br> If this is set to 0, then “sync_fifo_generic” module will be used for instantiation where a FIFO is needed.</br> If this is set to 1, then “sync_fifo_vendor” module will be used. </br> If planning to use this source, you need to generate a vendor FIFO IP in your environment and instantiate it under “sync_fifo_vendor” module.</br> In order to reduce the number of used resources/LUTs in a FPGA, the recommended way is using “sync_fifo_vendor” where you utilize built-in dedicated resources.|integer|    0 or 1|
|PAD_OPT|“zeros” : Assume zero valued pixels outside the image frame</br> “ones” : Assume 2<sup>DW_VD</sup>-1 valued pixels outside the image frame|string|    “zeros” or “ones”|

</br>

## Pinout Description

|Pin Name   |I/O       |Description        |Active State       |
|-          |-         |-                  |-                  |
|clk        |in        |Module clock       |rising edge        |
|rstb       |in        |Asynchronous reset |low                |
|**SLAVE VIDEO BUS**            
|s_vb_aux [3:0]|in    |Slave video bus auxiliary data</br> [0]: SOF</br> [1]: EOL</br> [2]: MSK</br>[3]: don’t care|high|
|s_vb_dat[DW_VD-1:0]|in    |Slave video bus pixel data|data|
|s_vb_val|in    |Data valid flag    |high|
|s_vb_rdy|out    |Ready to accept input data |high (handshake signal)|
|**MASTER VIDEO BUS**|
|m_vb_aux [3:0]    |out|    Master video bus auxiliary data</br> [0]: SOF </br>[1]: EOL </br>[2]: MSK</br> [3]: don’t care|    high|
|m_vb_dat[DW_VD-1:0]|    out    |Master video bus pixel data|data|
|m_vb_val|out    |Data valid flag    |high|
|m_vb_rdy|in    |Ready to accept input data    |high (handshake signal)|
|**SLAVE MEMORY BUS**|
|s_mb_adr [7:0]     |in     |Memory bus address         |data    |
|s_mb_wdt [15:0]    |in     |Memory bus write data      |data    |
|s_mb_rdt [15:0]    |out    |Memory bus read data       |data    |
|s_mb_val           |in     |Memory bus write valid     |high    |

</br>

## Interfaces
### VBUS Interface
This is the stream interface used in this IP. It consists of 4 signals/buses:
- DAT: data bus
- AUX: auxiliary data bus, includes SOF, EOL and MSK
- VAL: data valid bus (handshaking)
- RDY: ready to receive (handshaking)

In the following waveforms, how these signals are formed when a valid transaction is being performed is described.
A together assertion of VAL and RDY signals with the rising edge is a valid pixel transaction. The first pixel in the first line should be sent with SOF asserted. And the last pixel of every line should be sent with EOL asserted.
If filtering with mask is desired, MSK can be asserted with any pixel transaction.




<img src="doc/img/waveform1.png" height="150">
<img src="doc/img/waveform2.png" height="150"></br>

**Figure 2.** Input waveforms showing the start of a new frame/first line (top) and end of a line (bottom)
</br>


### MBUS Interface
This interface consists of 4 signals/buses and used to configure the memory of IP. The interface has the same signals a block memory interface has. So, it is very simple.

- ADR: memory address
- WDT: write data
- RDT: read data
- VAL: write enable


## Source Files
All source files are provided as text files coded in Verilog/SystemVerilog. The following table gives a brief explanation of each file.

|Source file            |Description                                |
|-                      |-                                          |
|**Design**             |                                           |
|median_filt.v          |Median filter top module                   |
|median_filt_buf.v      |Line buffering                             |
|median_filt_conf.v     |IP configuration memory                    |
|median_filt_core.v     |Finds median in a given matrix             |
|median_filt_filt.v     |Filtering logic with core                  |
|median_filt_tx.v       |Transmitter block with a buffer            |
|sync_fifo.v            |Instantiated FIFO used everywhere buffering needed|
|sync_fifo_generic.v    |Generic synchronized FIFO implementation|
|sync_fifo_vendor.v     |Template file for vendor FIFO IP instantiation |
|reset_gen.v            |Reset generation module                    |
|sort_NxN.v             |Sorting blocks                             |
|sort_N.v               |Sorting blocks                             |
|transpose_NxN.v        |Sorting blocks                             |
|dia_data_NxN.v         |Sorting blocks                             |
|single_pass_sorter.v   |Sorting blocks                             |        
|comparator.v           |Sorting blocks                             |
|**Simulation**         |                                           |
|tb.sv                  |Top testbench module                       |
|frame_timing_gen.v     |Video timing generation in TB              |
|pgm_reader.sv          |Read PGM image files into TB               |
|parallel_2_vbus.v      |Converts parallel video bus to VBUS        |
|vbus_2_parallel.v      |Converts VBUS to parallel video bus        |
|file_frame_grabber.sv  |Grabs frames from parallel video bus and record them to PGM files      |
|file_compare.sv        |Reads and compares two files               |
</br>


## Padding Options

<img src="doc/img/padopt.png" height="400"></br>

**Figure 3.** An example window placement is shown at the top edge of an image.</br> When `PAD_OPT = "zeros"`, outside pixel values are assumed to zero (left). </br>If `PAD_OPT = "ones"`, then these pixels will get the maximum value allowed (right) (e.g. 4095 for BPP=12)


## Simulation
In the **sim** folder, there is a "**.do**" file for automated simulations. You can use Modelsim to make simulations.
#### Modelsim
In order to run a simulation in Modelsim:
1. Run Modelsim
2. Change directory to "sim" folder with "cd" command
3. Enter "do sim.do"

The modelsim will run the simulation.

#### Other
If you want to use another simulator tool, .do file can be helpful


CHANGELOG.md
# Changelog
## 2D Median Filter with VBUS Interface
All notable changes to this IP will be documented in this file.

## 1.00 - 2020-12-16
- Initial release


sim/tb/tb.sv
`timescale 1ns/1ps

module tb ();

    // Testbench parameters
    parameter CLK_FREQ_HZ = 125_000_000;
    parameter CLK_PER_NS = 1_000_000_000 / CLK_FREQ_HZ; //Clock period in ns


    // Median Filter IP
    parameter BPP = 8;
    parameter SIZE = 3;
    parameter PAD_OPT = "zeros";
    parameter WIDTH = 640;
    parameter HEIGHT = 480;
    parameter IN_BUF_LEN = 4096;
    parameter OUT_BUF_LEN = 2048;
    parameter USE_VENDOR_FIFO = 0;


    // Video and memory bus widths
    parameter DW_VD = BPP;
    parameter DW_VX = 4;
    parameter DW_MA = 8;
    parameter DW_MD = 16;


    // File frame grabber
    parameter string OUTPUT_PATH = "../images/out/";
    parameter string FILE_PREFIX_INPUT = "../images/in/img";
    parameter string FILE_PREFIX_MASK = "../images/in/mask";
    parameter string FILE_EXT = ".pgm";
    parameter SCALETO8BIT = 0;
    parameter COLORED = 0;


    // Timing generation
    parameter T_IDLE2FVAL = 8192;
    parameter T_LVALHIGH_DVALHIGH = 16;
    parameter T_DVALLOW_LVALLOW = 16;
    parameter T_LVALLOW = 16;

    parameter FPS = 100;
    parameter TIMEOUT_MS = 1000;


    // Reference clock
    reg clk;

    // Async reset
    reg rstb;

    // Frame timing generator signals
    reg                 en_ftg;
    wire                fval_gen;
    wire                lval_gen;
    wire                dval_gen;
    reg                 reader_read;
    wire [$clog2(HEIGHT)-1:0] reader_row;
    wire [$clog2(WIDTH)-1:0] reader_col;


    // Reader data outputs
    wire [BPP-1:0]      pix_data_gen;
    wire [BPP-1:0]      mask_data_gen;


    // Mask always bit
    reg                 mask_always;
    wire                mask_data_sel;

    wire                error;

    wire                s_vb_val;
    wire                s_vb_rdy;
    wire [DW_VX-1:0]    s_vb_aux;
    wire [DW_VD-1:0]    s_vb_dat;

    wire                m_vb_val;
    wire                m_vb_rdy;
    wire  [DW_VX-1:0]   m_vb_aux;
    wire  [DW_VD-1:0]   m_vb_dat;

    wire                fval_out;
    wire                lval_out;
    wire                dval_out;
    wire [BPP-1:0]      pix_data_out;

    reg [DW_MA-1:0]     s_mb_adr;
    reg [DW_MD-1:0]     s_mb_wdt;
    wire [DW_MD-1:0]    s_mb_rdt;
    reg                 s_mb_val;


    // Input image file reader
    pgm_reader #(
        .FILE_PREFIX(FILE_PREFIX_INPUT),
        .FILE_EXT(FILE_EXT),
        .WIDTH(WIDTH),
        .HEIGHT(HEIGHT),
        .BPP(BPP))
    PGM_READER_IMAGE (
        .rstb(rstb),
        .clk(clk),
        .read(reader_read),
        .row(reader_row),
        .column(reader_col),
        .data_out(pix_data_gen)
    );


    // Mask image file reader
    pgm_reader #(
        .FILE_PREFIX(FILE_PREFIX_MASK),
        .FILE_EXT(FILE_EXT),
        .WIDTH(WIDTH),
        .HEIGHT(HEIGHT),
        .BPP(BPP))
    PGM_READER_MASK (
        .rstb(rstb),
        .clk(clk),
        .read(reader_read),
        .row(reader_row),
        .column(reader_col),
        .data_out(mask_data_gen)
    );


    // Video timing generation
    frame_timing_generator #(
        .CLK_PERIOD(CLK_PER_NS),
        .FPS(FPS), //frames per second
        .WIDTH(WIDTH), //Number of pixels in horizontal line
        .HEIGHT(HEIGHT), //Number of horizontal lines to be read in each frame
        .T_IDLE2FVAL(T_IDLE2FVAL),
        .T_LVALHIGH_DVALHIGH(T_LVALHIGH_DVALHIGH),
        .T_DVALLOW_LVALLOW(T_DVALLOW_LVALLOW),
        .T_LVALLOW(T_LVALLOW))
    FRAME_TIMING_GEN (
        .clk(clk), //Clock
        .rstb(rstb), //Active low reset
        .en(en_ftg), //Enable pulse that triggers the whole operation
        .fval(fval_gen), //Frame valid signal
        .lval(lval_gen), //Line valid signal
        .dval(dval_gen), //data valid signal
        .row(reader_row),
        .col(reader_col)
    );


    // FLDval to VBUS conversion
    parallel_2_vbus #(
        .BPP(BPP+1),
        .DW_VD(DW_VD),
        .DW_VX(DW_VX),
        .USE_VENDOR_FIFO(USE_VENDOR_FIFO))
    PAR_2_VBUS (
        .clk(clk),
        .rstb(rstb),

        .fval(fval_gen),
        .lval(lval_gen),
        .dval(dval_gen),
        .pix_data({mask_data_sel, pix_data_gen}),

        .m_vb_dat(s_vb_dat),
        .m_vb_val(s_vb_val),
        .m_vb_rdy(s_vb_rdy),
        .m_vb_aux(s_vb_aux)
    );


    // DUT
    median_filter #(
        .SIZE(SIZE),
        .PAD_OPT(PAD_OPT),
        .IN_BUF_LEN(IN_BUF_LEN),
        .OUT_BUF_LEN(OUT_BUF_LEN),
        .USE_VENDOR_FIFO(USE_VENDOR_FIFO),
        .DW_VD(DW_VD),
        .DW_VX(DW_VX),
        .DW_MA(DW_MA),
        .DW_MD(DW_MD))
    DUT (

        .clk(clk),
        .rstb(rstb),

        .error(error),

        // Slave video bus
        .s_vb_val(s_vb_val),
        .s_vb_rdy(s_vb_rdy),
        .s_vb_aux(s_vb_aux),
        .s_vb_dat(s_vb_dat),

        // Master video bus
        .m_vb_val(m_vb_val),
        .m_vb_rdy(m_vb_rdy),
        .m_vb_aux(m_vb_aux),
        .m_vb_dat(m_vb_dat),

        // Slave memory bus
        .s_mb_adr(s_mb_adr),
        .s_mb_wdt(s_mb_wdt),
        .s_mb_rdt(s_mb_rdt),
        .s_mb_val(s_mb_val)
    );


    // AXIS to FLDval conversion
    vbus_2_parallel #(
        .DW_VD(DW_VD),
        .DW_VX(DW_VX),
        .USE_VENDOR_FIFO(USE_VENDOR_FIFO),
        .MAX_HEIGHT(HEIGHT),
        .MAX_WIDTH(WIDTH),
        .BUF_LEN(OUT_BUF_LEN),
        .TIMEOUT_MS(TIMEOUT_MS),
        .CLK_FREQ_HZ(CLK_FREQ_HZ))
    VBUS_2_PAR (
        .clk(clk),
        .rstb(rstb),

        .s_vb_dat(m_vb_dat),
        .s_vb_val(m_vb_val),
        .s_vb_rdy(m_vb_rdy),
        .s_vb_aux(m_vb_aux),

        .fval(fval_out),
        .lval(lval_out),
        .dval(dval_out),
        .pix_data(pix_data_out)
    );


    // Grab output files and write it to PGM file
    file_frame_grabber #(
        .WIDTH(WIDTH),
        .HEIGHT(HEIGHT),
        .BPP(BPP),
        .OUTPUT_PATH(OUTPUT_PATH),
        .SCALETO8BIT(SCALETO8BIT),
        .COLORED(COLORED))
    FILE_FRAME_GRABBER(
        .rstb(rstb),
        .pix_clk(clk),
        .fval(fval_out),
        .lval(lval_out),
        .dval(dval_out),
        .pix_data(pix_data_out)
    );


    reg chk;
    wire res_1;
    wire res_2;
    file_compare #( .ISSUE_MESSAGES(0))
    FILE_COMP_1 (
        .file_path_1("../images/golden/filtered_w_mask.pgm"),
        .file_path_2("../images/out/_img0.pgm"),
        .check(chk),
        .result(res_1)
    );
    file_compare #( .ISSUE_MESSAGES(0))
    FILE_COMP_2 (
        .file_path_1("../images/golden/filtered.pgm"),
        .file_path_2("../images/out/_img2.pgm"),
        .check(chk),
        .result(res_2)
    );


    // Task for IP configuration
    task WRITE_TO_CONFIG;
        input [7:0] addr;
        input [15:0] data;
    begin
        s_mb_adr = addr;
        s_mb_wdt = data;
        s_mb_val = 1'b1;
        @(posedge clk);
        #2 s_mb_val = 1'b0;
        @(posedge clk);
    end
    endtask


    // Clock generation
    always #(CLK_PER_NS/2) clk = !clk;

    assign mask_data_sel = mask_always ? 1'b1 : mask_data_gen[0];

    initial begin
        rstb = 0;
        clk = 0;
        en_ftg = 0;
        reader_read = 0;
        s_mb_adr = 0;
        s_mb_wdt = 0;
        s_mb_val = 0;
        mask_always = 0;
        chk = 0;


        repeat (100) @(posedge clk);
        rstb = 1;
        repeat (100) @(posedge clk);

        // Read image/mask from file
        @(posedge clk);
        reader_read = 1;
        @(posedge clk);
        reader_read = 0;

        // Write custom image width value and see if the module resets blocks
        repeat (100) @(posedge clk);
        WRITE_TO_CONFIG(8'b01, 16'd640);    // Set WIDTH
        WRITE_TO_CONFIG(8'b10, 16'd480);    // Set HEIGHT
        WRITE_TO_CONFIG(8'b00, 16'd1);         // Enable the module



        en_ftg = 1;
        $display("### TB: Filtering image with mask (k=3)...");
        @(negedge fval_out);    // First frame

        mask_always = 1;
        $display("### TB: Filtering image without mask (k=3)...");
        @(negedge fval_out);    // Second frame

        // Skip 1 more frame because we assert "mask_always" after the second frame is started
        // The third frame will have "mask_always" correctly.
        @(negedge fval_out);     // Third frame
        repeat (100) @(posedge clk);

        $display("#############################################");
        $display("######## MEDIAN FILTER TEST RESULT ##########");
        // Check for tests
        @(posedge clk) chk = 1;
        @(posedge clk) chk = 0;
        if(res_1==0) $display("TEST 1 _____________________________ PASSED");
        else         $display("TEST 1 _____________________________ FAILED !!!");
        if(res_2==0) $display("TEST 2 _____________________________ PASSED");
        else         $display("TEST 2 _____________________________ FAILED !!!");
        $display("#############################################");

        @(posedge clk);
        $finish;

    end

endmodule
